mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,181 +165,76 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
         // The callee of the backprop Op
         func::FuncOp callee =
             SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
-        assert(callee);
-
-        // Creat constants
-        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(1));
+        assert(callee && "Expected a valid callee of type func.func");
 
         // Create mlir memref to llvm alloc
         StringRef allocFnName = "_mlir_memref_to_llvm_alloc";
-        Type allocFnSignature = LLVM::LLVMFunctionType::get(LLVM::LLVMPointerType::get(ctx),
-                                                            {IntegerType::get(ctx, 64)});
-        LLVM::LLVMFuncOp allocFnDecl =
-            ensureFunctionDeclaration(rewriter, op, allocFnName, allocFnSignature);
-        
+        Type allocFnSignature = LLVM::LLVMFunctionType::get(llvmPtrType, {rewriter.getI64Type()});
+        ensureFunctionDeclaration(rewriter, op, allocFnName, allocFnSignature);
+
         // Create mlir memref to llvm free
         StringRef freeFnName = "_mlir_memref_to_llvm_free";
```

Comment:
Can we get the freeFnName and allocFnName to be obtained in the same (or similar) way to how `--finalize-memref-to-llvm=use-generic-functions` uses it? Maybe we could even check this option, that way we only have to modify a single line to switch between free/malloc and `_mlir_memref_to_llvm` functions.

Reply:
Yes, this should be possible. We're actually exposing the same `use-generic-functions` option to the `gradient-to-llvm` pass, so I can look into reading it from this conversion pattern.